### PR TITLE
Add fee exemption whitelist

### PR DESCRIPTION
**Pull request by ivanzhelyazkov**
* Add fee exemption whitelist to Bancor V3 network contract, replacing the _bancorArbitrage immutable variable
* Any address in the fee exemption whitelist is exempt from trading and flashloan fees
* Admin can add and remove addresses from the whitelist
* Add tests
* Add deployment scripts

### CODE REVIEW

File: contracts/network/BancorNetwork.sol [hunk 1]
```diff
@@ -692,7 +702,10 @@ contract BancorNetwork is IBancorNetwork, Upgradeable, ReentrancyGuardUpgradeabl
         uint256 minReturnAmount,
         uint256 deadline,
         address beneficiary
-    ) external payable whenNotPaused only(_bancorArbitrage) returns (uint256) {
+    ) external payable whenNotPaused returns (uint256) {
+        if (!_feeExemptionWhitelist.contains(msg.sender)) {
+            revert AccessDenied();
+        }
```
Comment (lbeder) [inline contracts/network/BancorNetwork.sol]: Missing new line?

File: contracts/network/BancorNetwork.sol [hunk 2]
```diff
@@ -715,7 +729,10 @@ contract BancorNetwork is IBancorNetwork, Upgradeable, ReentrancyGuardUpgradeabl
         uint256 maxSourceAmount,
         uint256 deadline,
         address beneficiary
-    ) external payable whenNotPaused only(_bancorArbitrage) returns (uint256) {
+    ) external payable whenNotPaused returns (uint256) {
+        if (!_feeExemptionWhitelist.contains(msg.sender)) {
+            revert AccessDenied();
+        }
```
Comment (lbeder) [inline contracts/network/BancorNetwork.sol]: Perhaps this can be refactored to a modifier?

File: contracts/network/BancorNetwork.sol [hunk 3]
```diff
@@ -878,6 +896,77 @@ contract BancorNetwork is IBancorNetwork, Upgradeable, ReentrancyGuardUpgradeabl
         _depositingEnabled = status;
     }
 
+    /**
+     * @inheritdoc IBancorNetwork
+     */
+    function feeExemptionWhitelist() external view returns (address[] memory) {
+        uint256 length = _feeExemptionWhitelist.length();
+        address[] memory list = new address[](length);
+        for (uint256 i = 0; i < length; i++) {
```
Comment (lbeder) [inline contracts/network/BancorNetwork.sol]: Nitpick: ++i

File: contracts/network/BancorNetwork.sol [hunk 4]
```diff
@@ -878,6 +896,77 @@ contract BancorNetwork is IBancorNetwork, Upgradeable, ReentrancyGuardUpgradeabl
...
+     * @dev adds multiple addresses to the fee exemption whitelist
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function addAddressesToWhitelist(address[] calldata addrs) external onlyAdmin {
+        uint256 length = addrs.length;
+
+        for (uint256 i = 0; i < length; i++) {
```
Comment (lbeder) [inline contracts/network/BancorNetwork.sol]: Nitpick: ++i

File: contracts/network/BancorNetwork.sol [hunk 5]
```diff
@@ -878,6 +896,77 @@ contract BancorNetwork is IBancorNetwork, Upgradeable, ReentrancyGuardUpgradeabl
...
+            _addToWhitelist(addrs[i]);
+        }
+    }
+
+    /**
+     * @dev adds an address to the fee exemption whitelist
+     */
+    function _addToWhitelist(address addr) private validExternalAddress(addr) {
```
Comment (lbeder) [inline contracts/network/BancorNetwork.sol]: Solidity official style guide recommends putting all the internal and private functions after all the public and external functions

**Review by lbeder — CHANGES_REQUESTED** [review]
Please also note the solhint warnings (which should be either fixed or ignored via lint rules)

File: contracts/network/BancorNetwork.sol (hunk 1 above)
Comment (ivanzhelyazkov) [reply contracts/network/BancorNetwork.sol]: replaced with modifier

File: contracts/network/BancorNetwork.sol (hunk 2 above)
Comment (ivanzhelyazkov) [reply contracts/network/BancorNetwork.sol]: refactored

File: contracts/network/BancorNetwork.sol (hunk 3 above)
Comment (ivanzhelyazkov) [reply contracts/network/BancorNetwork.sol]: fixed

File: contracts/network/BancorNetwork.sol (hunk 4 above)
Comment (ivanzhelyazkov) [reply contracts/network/BancorNetwork.sol]: fixed

File: contracts/network/BancorNetwork.sol (hunk 5 above)
Comment (ivanzhelyazkov) [reply contracts/network/BancorNetwork.sol]: fixed

**Review by lbeder — DISMISSED** [review]

File: hardhat.config.ts [hunk 6]
```diff
@@ -124,7 +124,7 @@ const config: HardhatUserConfig = {
                 settings: {
                     optimizer: {
                         enabled: true,
-                        runs: 200
+                        runs: 180
```
Comment (yudilevi) [inline hardhat.config.ts]: contract size issue? this might have some unexpected implications

Comment (ivanzhelyazkov) [reply hardhat.config.ts]: Yeah, it’s over the contract size without reducing the optimizer runs

Comment (yudilevi) [reply hardhat.config.ts]: We can remove the method that allows whitelisting multiple addresses at once

Comment (ivanzhelyazkov) [reply hardhat.config.ts]: removed

**Review by yudilevi — APPROVED** [review]

**Pull request merged**